### PR TITLE
C++: Fix `strtol` model

### PR DESCRIPTION
**Pull request by MathiasVP**
We had incorrectly included `strtol` in our model of "pure funtions". This meant that for a call such as:
```cpp
void test(const char* s) {
  char* end;
  strtod(s, &end);
}
```
we didn't model the fact that `end` was written to by `strtod`. I'm not sure if this would have any real query result changes, but it made an impact on a query I was playing with locally.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  Strtol() { this.hasGlobalOrStdOrBslName(strtol()) }
+
+  override predicate hasArrayInput(int bufParam) {
+    // All the functions given by `strtol()` takes a `const char*` input as the first parameter
+    bufParam = 0
+  }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) { bufParam = 0 }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
```
Comment (alexet) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll:21]: While we are writing this, do we want flow from the first argument to the dereference of the second argument as well?

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    (
+      input.isParameter(0)
+      or
+      input.isParameterDeref(0)
+    ) and
+    output.isReturnValue()
+  }
+
+  override predicate parameterNeverEscapes(int i) {
+    this.getParameter(i).getUnspecifiedType() instanceof PointerType
```
Comment (alexet) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll:31]: Parameter `0` does escape into parameter `1`. Although we have no way of writing that.

**Review by alexet — CHANGES_REQUESTED** [review]

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll:31]: Indeed, we don't have a way to specify this, no. I can write it as a comment, though 👍

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll:21]: We can do that for sure, yeah 👍

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strtol.qll]: Fixed in 866c0039a7af0ead9d2ecce665c8a468a0f32731. I also simplified the code a bit in the process.

**Comment by MathiasVP** [comment]
@alexet I believe I've addressed all your comments now 🤞

**Review by alexet — APPROVED** [review]
LGTM, Assuming DCA is good.

**Comment by MathiasVP** [comment]
DCA looks good. A couple of FPs on `cpp/redundant-null-check-simple` has been removed because we now realize that the second parameter is an output parameter 🎉

**Pull request merged**